File: payload/ariadne-node/enrich.mjs
#!/usr/bin/env node
/**
 * AEGIS enrich, the semantic layer: applies an LLM on top of the code+docs
 * graph to produce cached, hash-keyed insights (module intent, hotspot-file
 * summaries, architecture observations).
 *
 * Token economics: each target is summarized ONCE per content-hash. Re-runs
 * (CI on merge, local) only pay for what actually changed. All prompts are
 * built from the graph (outlines, callers, topics, tables, endpoints) not
 * raw file dumps, keeping each prompt ~1-2 KB.
 *
 * Providers (env-selected, explicit opt-in, see PRIVACY.md):
 *   ANTHROPIC_API_KEY          -> Anthropic API (model: AEGIS_MODEL or claude-haiku-4-5)
 *   OPENAI_API_KEY [+ OPENAI_BASE_URL] -> any OpenAI-compatible endpoint,
 *       including LOCAL servers (Ollama/llama.cpp/vLLM at http://localhost:11434/v1)
 *       for fully offline enrichment
 *   --provider mock            -> deterministic canned output (self-tests)
 *
 * Usage: node .ariadne/enrich.mjs [--limit 20] [--provider anthropic|openai|mock]
 * Outputs: insights table in index.db (+ staleness tracking), docs/generated/insights.md,
 * and the `explain` MCP tool serves them to agents.
 */
import Database from "better-sqlite3";
import fs from "node:fs";
import path from "node:path";
import { createHash } from "node:crypto";

const ROOT = process.cwd();
const DB_PATH = path.join(process.env.ARIADNE_HOME ?? ROOT, ".ariadne", "index.db");
const args = process.argv.slice(2);
const argOf = (k, d) => { const i = args.indexOf(k); return i >= 0 ? args[i + 1] : d; };
const LIMIT = parseInt(argOf("--limit", "20"), 10);
let PROVIDER = argOf("--provider",
  process.env.ANTHROPIC_API_KEY ? "anthropic" : process.env.OPENAI_API_KEY ? "openai" : null);
const PLAN = args.includes("--plan");
const APPLY = argOf("--apply", null);

if (!fs.existsSync(DB_PATH)) { console.error("Index not found, run the indexer first."); process.exit(1); }
if (!PROVIDER && !PLAN && !APPLY) {
  console.error(`No LLM provider configured. Set ANTHROPIC_API_KEY, or OPENAI_API_KEY
(+ OPENAI_BASE_URL for local servers like Ollama: http://localhost:11434/v1),
or pass --provider mock. Enrichment is opt-in and off by default, see PRIVACY.md.`);
  process.exit(2);
}

const db = new Database(DB_PATH);
db.pragma("busy_timeout = 10000");
db.exec(`CREATE TABLE IF NOT EXISTS insights(
  target TEXT PRIMARY KEY, kind TEXT, hash TEXT, summary TEXT,
  model TEXT, generated_at REAL)`);
const q = (sql, ...a) => db.prepare(sql).all(...a);
const svc = (p) => p.split("/")[0];
// insights describe production intent: seam facts from test files stay out of
// the prompts (column exists only once an indexer >= schema v5 has run)
const PROD = q("SELECT COUNT(*) c FROM pragma_table_info('files') WHERE name='is_test'")[0].c ? " AND f.is_test=0" : "";

/* ---------------- LLM call ---------------- */
async function complete(prompt) {
  if (PROVIDER === "mock") {
    return `[mock insight] ${prompt.slice(prompt.indexOf("TARGET:"), prompt.indexOf("TARGET:") + 60)}…, intent summarized deterministically for self-tests.`;
  }
  if (PROVIDER === "anthropic") {
    const res = await fetch("https://api.anthropic.com/v1/messages", {
      method: "POST",
      headers: { "content-type": "application/json", "x-api-key": process.env.ANTHROPIC_API_KEY,
                 "anthropic-version": "2023-06-01" },
      body: JSON.stringify({ model: process.env.AEGIS_MODEL ?? "claude-haiku-4-5",
        max_tokens: 400, messages: [{ role: "user", content: prompt }] }),
    });
    if (!res.ok) throw new Error(`anthropic ${res.status}: ${(await res.text()).slice(0, 200)}`);
    const j = await res.json();
    return j.content.map((c) => c.text ?? "").join("");
  }
  // openai-compatible (includes local Ollama/llama.cpp/vLLM via OPENAI_BASE_URL)
  const base = (process.env.OPENAI_BASE_URL ?? "https://api.openai.com/v1").replace(/\/$/, "");
  const res = await fetch(`${base}/chat/completions`, {
    method: "POST",
    headers: { "content-type": "application/json", authorization: `Bearer ${process.env.OPENAI_API_KEY ?? "local"}` },
    body: JSON.stringify({ model: process.env.AEGIS_MODEL ?? "gpt-4o-mini",
      max_tokens: 400, messages: [{ role: "user", content: prompt }] }),
  });
  if (!res.ok) throw new Error(`openai ${res.status}: ${(await res.text()).slice(0, 200)}`);
  return (await res.json()).choices[0].message.content;
}

/* ---------------- targets: modules + hotspot files, hash-keyed ---------------- */
function moduleTargets() {
  const files = q("SELECT path, hash FROM files");
  const mods = new Map();
  for (const f of files) {
    const m = mods.get(svc(f.path)) ?? { paths: [], hashes: [] };
    m.paths.push(f.path); m.hashes.push(f.hash ?? "");
    mods.set(svc(f.path), m);
  }
  return [...mods.entries()].map(([name, m]) => ({
    kind: "module", target: name,
    hash: createHash("sha1").update(m.hashes.sort().join("|")).digest("hex"),
  }));
}
function hotspotTargets(n = 12) {
  return q(`SELECT f.path, f.hash, COUNT(e.src) deps FROM files f JOIN edges e ON e.dst=f.id
            GROUP BY f.id ORDER BY deps DESC LIMIT ?`, n)
    .map((r) => ({ kind: "file", target: r.path, hash: r.hash ?? "" }));
}

/* ---------------- prompt packs from the graph (no raw file dumps) ---------------- */
function promptFor(t) {
  const header = `You are enriching a codebase knowledge graph. Write a dense, factual summary (5-8 sentences):
purpose/intent, key responsibilities, how it connects to the rest of the system, and any invariants
or gotchas a developer must know before changing it. No preamble, no markdown headers.

TARGET: ${t.kind} ${t.target}\n`;
  if (t.kind === "module") {
    const syms = q(`SELECT s.name, s.kind FROM symbols s JOIN files f ON f.id=s.file_id
                    WHERE f.path LIKE ? AND s.kind IN ('class','type') LIMIT 25`, t.target + "/%");
    const topics = q(`SELECT DISTINCT m.topic, m.direction FROM msg_edges m JOIN files f ON f.id=m.file_id WHERE f.path LIKE ?${PROD}`, t.target + "/%");
    const tables = q(`SELECT DISTINCT a.tbl, a.mode FROM db_access a JOIN files f ON f.id=a.file_id WHERE f.path LIKE ?${PROD}`, t.target + "/%");
    const eps = q(`SELECT e.method, e.path FROM http_endpoints e JOIN files f ON f.id=e.file_id WHERE f.path LIKE ?${PROD} LIMIT 15`, t.target + "/%");
    return header +
      `Classes/types: ${syms.map((s) => s.name).join(", ") || "n/a"}\n` +
      `Kafka: ${topics.map((x) => `${x.direction} ${x.topic}`).join(", ") || "none"}\n` +
      `DB tables: ${tables.map((x) => `${x.tbl}(${x.mode})`).join(", ") || "none"}\n` +
      `HTTP endpoints: ${eps.map((e) => `${e.method} ${e.path}`).join(", ") || "none"}\n`;
  }
  const outline = q(`SELECT s.name, s.kind, s.parent FROM symbols s JOIN files f ON f.id=s.file_id
                     WHERE f.path=? ORDER BY s.line LIMIT 40`, t.target);
  const callers = q(`SELECT COUNT(*) c FROM edges e JOIN files f ON f.id=e.dst WHERE f.path=?`, t.target)[0]?.c ?? 0;
  return header +
    `Outline: ${outline.map((s) => (s.parent ? s.parent + "." : "") + s.name + ":" + s.kind).join(", ")}\n` +
    `Dependent files: ${callers} (high blast radius)\n`;
}

/* ---------------- plan/apply: external drivers (e.g. Copilot via the VS Code extension) ---------------- */
const allTargets = [...moduleTargets(), ...hotspotTargets()];
if (PLAN) {
  const get0 = db.prepare("SELECT hash FROM insights WHERE target=?");
  const plan = allTargets
    .filter((t) => { try { const p = get0.get(t.target); return !p || p.hash !== t.hash; } catch { return true; } })
    .slice(0, LIMIT)
    .map((t) => ({ ...t, prompt: promptFor(t) }));
  console.log(JSON.stringify(plan));
  process.exit(0);
}
if (APPLY) {
  const items = JSON.parse(fs.readFileSync(APPLY, "utf8")); // [{target,kind,hash,summary,model}]
  const putA = db.prepare(`INSERT OR REPLACE INTO insights(target, kind, hash, summary, model, generated_at)
                           VALUES(?,?,?,?,?,?)`);
  for (const it of items) putA.run(it.target, it.kind, it.hash, String(it.summary).slice(0, 4000), it.model ?? "external", Date.now() / 1000);
  console.log(`Applied ${items.length} insights.`);
  writeInsightsMd();
  process.exit(0);
}

/* ---------------- run ---------------- */
const targets = allTargets;
const get = db.prepare("SELECT hash FROM insights WHERE target=?");
const put = db.prepare(`INSERT OR REPLACE INTO insights(target, kind, hash, summary, model, generated_at)
                        VALUES(?,?,?,?,?,?)`);
let fresh = 0, cached = 0, failed = 0;
for (const t of targets) {
  if (fresh >= LIMIT) break;
  const prev = get.get(t.target);
  if (prev && prev.hash === t.hash) { cached++; continue; }
  try {
    const summary = (await complete(promptFor(t))).trim();
    put.run(t.target, t.kind, t.hash, summary, PROVIDER === "mock" ? "mock" : (process.env.AEGIS_MODEL ?? PROVIDER), Date.now() / 1000);
    fresh++;
    console.log(`  + ${t.kind}: ${t.target}`);
  } catch (e) { failed++; console.error(`  ! ${t.target}: ${e.message}`); }
}
console.log(`Enrichment: ${fresh} generated, ${cached} cached (hash-unchanged), ${failed} failed.`);

/* ---------------- insights.md ---------------- */
writeInsightsMd();
function writeInsightsMd() {
const rows = q("SELECT target, kind, summary, model, generated_at FROM insights ORDER BY kind, target");
if (rows.length) {
  const out = path.join(ROOT, "docs", "generated");
  fs.mkdirSync(out, { recursive: true });
  let md = `<!-- generated by aegis enrich, cached by content hash; do not edit -->\n\n# Semantic Insights (LLM-enriched)\n\n`;
  md += `_Each entry is regenerated only when its content hash changes. Provider/model per entry._\n\n`;
  for (const r of rows.filter((x) => x.kind === "module")) md += `## Module: ${r.target}\n${r.summary}\n\n`;
  const hot = rows.filter((x) => x.kind === "file");
  if (hot.length) md += `## High-blast-radius files\n\n` + hot.map((r) => `**\`${r.target}\`**, ${r.summary}`).join("\n\n") + "\n";
  fs.writeFileSync(path.join(out, "insights.md"), md);
  console.log("  + docs/generated/insights.md");
}
}
db.close();

File: payload/ariadne-python/enrich.py
#!/usr/bin/env python3
"""AEGIS enrich (Python edition). LLM semantic layer over the graph.
Hash-cached insights per module/hotspot file; providers: ANTHROPIC_API_KEY,
OPENAI_API_KEY (+OPENAI_BASE_URL for local Ollama/vLLM), or --provider mock."""
import argparse
import hashlib
import json
import os
import sqlite3
import sys
import time
import urllib.request
from pathlib import Path

ROOT = Path.cwd()
DB = Path(os.environ.get("ARIADNE_HOME", ROOT)) / ".ariadne" / "index.db"

ap = argparse.ArgumentParser()
ap.add_argument("--limit", type=int, default=20)
ap.add_argument("--plan", action="store_true")
ap.add_argument("--apply")
ap.add_argument("--provider", choices=["anthropic", "openai", "mock"],
                default=("anthropic" if os.environ.get("ANTHROPIC_API_KEY")
                         else "openai" if os.environ.get("OPENAI_API_KEY") else None))
A = ap.parse_args()
if not DB.exists():
    sys.exit("Index not found, run the indexer first.")
if not A.provider and not A.plan and not A.apply:
    sys.exit("No LLM provider configured. Set ANTHROPIC_API_KEY, or OPENAI_API_KEY "
             "(+OPENAI_BASE_URL for local servers like Ollama), or pass --provider mock. "
             "Enrichment is opt-in, see PRIVACY.md.")

con = sqlite3.connect(DB)
con.row_factory = sqlite3.Row
con.execute("""CREATE TABLE IF NOT EXISTS insights(
  target TEXT PRIMARY KEY, kind TEXT, hash TEXT, summary TEXT,
  model TEXT, generated_at REAL)""")
q = lambda sql, *a: con.execute(sql, a).fetchall()
svc = lambda p: p.split("/")[0]


def _post(url, headers, body):
    req = urllib.request.Request(url, data=json.dumps(body).encode(),
                                 headers={"content-type": "application/json", **headers})
    with urllib.request.urlopen(req, timeout=120) as r:
        return json.loads(r.read())


def complete(prompt):
    if A.provider == "mock":
        i = prompt.index("TARGET:")
        return f"[mock insight] {prompt[i:i+60]}…, intent summarized deterministically for self-tests."
    model = os.environ.get("AEGIS_MODEL")
    if A.provider == "anthropic":
        j = _post("https://api.anthropic.com/v1/messages",
                  {"x-api-key": os.environ["ANTHROPIC_API_KEY"], "anthropic-version": "2023-06-01"},
                  {"model": model or "claude-haiku-4-5", "max_tokens": 400,
                   "messages": [{"role": "user", "content": prompt}]})
        return "".join(c.get("text", "") for c in j["content"])
    base = os.environ.get("OPENAI_BASE_URL", "https://api.openai.com/v1").rstrip("/")
    j = _post(f"{base}/chat/completions",
              {"authorization": f"Bearer {os.environ.get('OPENAI_API_KEY', 'local')}"},
              {"model": model or "gpt-4o-mini", "max_tokens": 400,
               "messages": [{"role": "user", "content": prompt}]})
    return j["choices"][0]["message"]["content"]


def module_targets():
    mods = {}
    for f in q("SELECT path, hash FROM files"):
        mods.setdefault(svc(f["path"]), []).append(f["hash"] or "")
    return [{"kind": "module", "target": name,
             "hash": hashlib.sha1("|".join(sorted(hs)).encode()).hexdigest()}
            for name, hs in mods.items()]


def hotspot_targets(n=12):
    return [{"kind": "file", "target": r["path"], "hash": r["hash"] or ""}
            for r in q("SELECT f.path, f.hash, COUNT(e.src) d FROM files f "
                       "JOIN edges e ON e.dst=f.id GROUP BY f.id ORDER BY d DESC LIMIT ?", n)]


def prompt_for(t):
    header = ("You are enriching a codebase knowledge graph. Write a dense, factual summary "
              "(5-8 sentences): purpose/intent, key responsibilities, how it connects to the rest "
              "of the system, and any invariants or gotchas a developer must know before changing it. "
              f"No preamble, no markdown headers.\n\nTARGET: {t['kind']} {t['target']}\n")
    if t["kind"] == "module":
        like = t["target"] + "/%"
        syms = q("SELECT s.name FROM symbols s JOIN files f ON f.id=s.file_id "
                 "WHERE f.path LIKE ? AND s.kind IN ('class','type') LIMIT 25", like)
        topics = q("SELECT DISTINCT m.topic, m.direction FROM msg_edges m JOIN files f ON f.id=m.file_id WHERE f.path LIKE ?", like)
        tables = q("SELECT DISTINCT a.tbl, a.mode FROM db_access a JOIN files f ON f.id=a.file_id WHERE f.path LIKE ?", like)
        eps = q("SELECT e.method, e.path FROM http_endpoints e JOIN files f ON f.id=e.file_id WHERE f.path LIKE ? LIMIT 15", like)
        return header + (
            f"Classes/types: {', '.join(s['name'] for s in syms) or 'n/a'}\n"
            f"Kafka: {', '.join(x['direction'] + ' ' + x['topic'] for x in topics) or 'none'}\n"
            f"DB tables: {', '.join(x['tbl'] + '(' + x['mode'] + ')' for x in tables) or 'none'}\n"
            f"HTTP endpoints: {', '.join(e['method'] + ' ' + e['path'] for e in eps) or 'none'}\n")
    outline = q("SELECT s.name, s.kind, s.parent FROM symbols s JOIN files f ON f.id=s.file_id "
                "WHERE f.path=? ORDER BY s.line LIMIT 40", t["target"])
    callers = q("SELECT COUNT(*) c FROM edges e JOIN files f ON f.id=e.dst WHERE f.path=?", t["target"])[0]["c"]
    return header + (
        "Outline: " + ", ".join((s["parent"] + "." if s["parent"] else "") + s["name"] + ":" + s["kind"] for s in outline)
        + f"\nDependent files: {callers} (high blast radius)\n")


targets = module_targets() + hotspot_targets()

if A.plan:
    plan = []
    for t in targets:
        prev = con.execute("SELECT hash FROM insights WHERE target=?", (t["target"],)).fetchone()
        if prev and prev["hash"] == t["hash"]:
            continue
        plan.append({**t, "prompt": prompt_for(t)})
        if len(plan) >= A.limit:
            break
    print(json.dumps(plan))
    sys.exit(0)

if A.apply:
    items = json.loads(Path(A.apply).read_text(encoding="utf-8"))
    for it in items:
        con.execute("INSERT OR REPLACE INTO insights(target, kind, hash, summary, model, generated_at) VALUES(?,?,?,?,?,?)",
                    (it["target"], it["kind"], it["hash"], str(it["summary"])[:4000], it.get("model", "external"), time.time()))
    con.commit()
    print(f"Applied {len(items)} insights.")

fresh = cached = failed = 0
if A.apply:
    targets = []
for t in targets:
    if fresh >= A.limit:
        break
    prev = con.execute("SELECT hash FROM insights WHERE target=?", (t["target"],)).fetchone()
    if prev and prev["hash"] == t["hash"]:
        cached += 1
        continue
    try:
        summary = complete(prompt_for(t)).strip()
        con.execute("INSERT OR REPLACE INTO insights(target, kind, hash, summary, model, generated_at) VALUES(?,?,?,?,?,?)",
                    (t["target"], t["kind"], t["hash"], summary,
                     "mock" if A.provider == "mock" else os.environ.get("AEGIS_MODEL", A.provider), time.time()))
        fresh += 1
        print(f"  + {t['kind']}: {t['target']}")
    except Exception as e:  # noqa: BLE001
        failed += 1
        print(f"  ! {t['target']}: {e}", file=sys.stderr)
con.commit()
print(f"Enrichment: {fresh} generated, {cached} cached (hash-unchanged), {failed} failed.")

rows = q("SELECT target, kind, summary FROM insights ORDER BY kind, target")
if rows:
    out = ROOT / "docs" / "generated"
    out.mkdir(parents=True, exist_ok=True)
    md = ("<!-- generated by aegis enrich, cached by content hash; do not edit -->\n\n"
          "# Semantic Insights (LLM-enriched)\n\n"
          "_Each entry is regenerated only when its content hash changes._\n\n")
    for r in rows:
        if r["kind"] == "module":
            md += f"## Module: {r['target']}\n{r['summary']}\n\n"
    hot = [r for r in rows if r["kind"] == "file"]
    if hot:
        md += "## High-blast-radius files\n\n" + "\n\n".join(f"**`{r['target']}`**, {r['summary']}" for r in hot) + "\n"
    (out / "insights.md").write_text(md, encoding="utf-8")
    print("  + docs/generated/insights.md")
